modules/control/spectate.lua

```diff
@@ -0,0 +1,168 @@
+    if character and character.valid then
+        local opened = player.opened
+        player.teleport(character.position, character.surface)
+        player.set_controller{ type = defines.controllers.character, character = character }
+        if opened then player.opened = opened end -- Maintain opened after controller change
+    else
+        player.ticks_to_respawn = 300
+    end
+end
+
+--- Test if a player is in follow mode
+-- @tparam LuaPlayer player The player to test the follow mode of
+-- @treturn boolean True if the player is in follow mode
+function Public.is_following(player)
+    assert(player and player.valid, 'Invalid player given to follower')
+    return following[player.index] ~= nil
+end
+
+--- Puts a player into spectator and follows an entity as it moves
+-- @tparam LuaPlayer player The player that will follow the entity
+-- @tparam ?LuaPlayer|LuaEntity entity The player or entity that will be followed
+function Public.start_follow(player, entity)
+    assert(player and player.valid, 'Invalid player given to follower')
+    assert(entity and entity.valid, 'Invalid entity given to follower')
+    local spectate = Public.start_spectate(player)
+
+    player.close_map()
+    follow_label(player.gui.screen, entity)
+    player.teleport(entity.position, entity.surface)
+    following[player.index] = { player, entity, entity.position, spectate }
+end
+
+--- Stops a player from following their current entity, can also return a player from spectate
+-- @tparam LuaPlayer player The player that you want to stop following their entity
+function Public.stop_follow(player)
+    assert(player and player.valid, 'Invalid player given to follower')
+    if following[player.index] and following[player.index][4] then Public.stop_spectate(player) end
+
+    Gui.destroy_if_valid(player.gui.screen[follow_label.name])
+    following[player.index] = nil
+end
+
+--- Used to stop all players following an entity or player
+function Public.stop_all()
+    for key, data in pairs(following) do
+        Public.stop_follow(data[1])
+    end
+end
+
+----- Gui -----
+
+--- Label used to show that the player is following, also used to allow esc to stop following
+-- @element follow_label
+follow_label =
+Gui.element(function(event_trigger, parent, target)
+    Gui.destroy_if_valid(parent[event_trigger])
+
+    local label = parent.add{
+        name = event_trigger,
+        type = 'label',
+        style = 'heading_1_label',
+        caption = 'Following '..target.name..'.\nClick here or press esc to stop following.'
+    }
+
+    local player = Gui.get_player_from_element(parent)
+    local res = player.display_resolution
+    label.location = {0, res.height-150}
+    label.style.width = res.width
+    label.style.horizontal_align = 'center'
+    player.opened = label
+
+    return label
+end)
+:on_click(Public.stop_follow)
+:on_close(function(player)
+    -- Dont call set_controller during on_close as it invalidates the controller
```

Comment:
```suggestion
    -- Don't call set_controller during on_close as it invalidates the controller
```
or `Do not` if you're afraid of apostrophes :P

Reply:
> or Do not if you're afraid of apostrophes :P

Absolutely not, they are far better than speach marks. :)